waspc/src/Wasp/AI/GenerateNewProject/LogMsg.hs

```diff
@@ -0,0 +1,55 @@
+module Wasp.AI.GenerateNewProject.LogMsg
+  ( LogMsg,
+    styled,
+    toPlainString,
+    toTermString,
+    fromText,
+    Style (..),
+  )
+where
+
+import Data.String (IsString (fromString))
+import qualified Data.Text as T
+import qualified Wasp.Util.Terminal as Term
+
+data LogMsg = Plain String | Concat [LogMsg] | Styled Style LogMsg
+  deriving (Show, Eq)
+
+data Style = Important | Generating | Fixing | Error | Custom [Term.Style]
+  deriving (Show, Eq)
+
+instance IsString LogMsg where
+  fromString = Plain
+
+fromText :: T.Text -> LogMsg
+fromText = fromString . T.unpack
+
+instance Semigroup LogMsg where
+  Concat ms1 <> Concat ms2 = Concat $ ms1 <> ms2
+  Concat ms1 <> ms2 = Concat $ ms1 <> [ms2]
+  ms1 <> Concat ms2 = Concat $ ms1 : ms2
+  m1 <> m2 = Concat [m1, m2]
+
+instance Monoid LogMsg where
+  mempty = Plain ""
+
+styled :: Style -> LogMsg -> LogMsg
+styled s m = Styled s m
+
+toPlainString :: LogMsg -> String
+toPlainString (Plain m) = m
+toPlainString (Styled _ m) = toPlainString m
+toPlainString (Concat ms) = concat $ toPlainString <$> ms
+
+toTermString :: LogMsg -> String
+toTermString (Plain m) = m
+toTermString (Styled s m) =
+  Term.applyStyles (styleToTermStyles s) $ toTermString m
+  where
+    styleToTermStyles = \case
+      Important -> [Term.Bold, Term.Magenta]
+      Generating -> [Term.Blue]
+      Fixing -> [Term.Green]
+      Error -> [Term.Red]
+      Custom styles -> styles
+toTermString (Concat ms) = concat $ toTermString <$> ms
```

Comment:
`map` would make this more readable on first read 👍

Reply:
I find this more readable! Experience. But I suggest trying to get into it, I think once you get used to it, code gets easier to read. `<$>` is very common in Haskell, and you should get used to being able to read it quickly.

One thing I want to emphasize is that in Haskell, I most often don't bother understanding what every symbol does. I know compiler has my back and that stuff is "connected" correctly. What I care about is the general flow of data and what is happening.

So for example, when reading this code, I read it like this:
> ms is taken out of Concat, then toTermString is applied to it, and then concat is applied to that.

Notice that I don't actually care too much, at least in the first read, if it is `<$>` or `$`, I see it all as application. It doesn't really matter to me, I know that the thing on the right is now "term stringified" and that at the end all stuff is concatenated.

And when reading code like that, it is easier to read it when using `$` and `<$>` then when using `fmap` or `map` because they break my reading flow and force me to think about them.

Same is if I encounter an expression with a bunch of `>>=`, `<$>`, `$`, `>>`, `&`, ... .
Most of the time, I just need a general idea of what it does, and for that I don't need to understand exactly why each of those operators was used in that specific place. If I need to modify it, then I will dig a bit deeper, and again compiler will be helping me out.

Reply:
I still think that `map` is more easier to read for developers that have < 2 years of Haskell experience. But as we talked about it IRL, this seems a matter of personal preference and as I said already, I won't press this matter any more.